File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        n = len(height)
        
        left, right = 0, n-1
        water = 0
        
        
        while left < right:
            width = right - left
            height_min = min(height[left], height[right])
            current_water = width * height_min
            
            water = max(water, current_water)
            
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return water
        

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        
        int n = height.size();
        int left = 0, right = n - 1;
        int water = 0;
        
        while (left < right) {
            int width = right - left;
            int minHeight = min(height[left], height[right]);
            int curWater = minHeight * width;
            
            water = max(water, curWater);
            
            if (height[left] < height[right]){
                left++;
            } else {
                right--;
            }
            
        }
        
        return water;
    }
};